App/IntentRunner.swift

```
@@ -11,22 +11,41 @@ import UIKit
 import MediaPlayer
 
 enum IntentRunner {
+    private static var activePlayer: [UUID: AudioPlayer] = [:]
+
     static func perform(intent: SoundIntent) async throws -> some IntentResult {
+        let soundID = intent.sound.id
+
+        if let player = activePlayer[soundID], player.isPlaying {
+            print("Stopping sound \(intent.sound.title)")
+            player.stop()
+            activePlayer[soundID] = nil
+            return .result()
+        }
+
+        guard let fileURL = intent.sound.file.fileURL, let newPlayer = try? AudioPlayer(url: fileURL) else {
+            print("Failed to create player for \(intent.sound.title)")
+            return .result()
+        }
         print("Start playing \(intent.sound.title)")
-        let player = try! AudioPlayer(url: intent.sound.file.fileURL!)
 
         print("Created sound for \(intent.sound.title)")
 
-        if intent.isFullBlast && player.isOnSpeaker {
+        if intent.isFullBlast && newPlayer.isOnSpeaker {
             await MPVolumeView.setVolume(1)
         }
 
+        activePlayer[soundID] = newPlayer
+
         do {
             print("Playing sound \(intent.sound.title)")
-            try await player.playOnQueue()
+            try await newPlayer.playOnQueue()
             print("Sound \(intent.sound.title) stopped")
+
+            activePlayer[intent.sound.id] = nil
         } catch {
-            print("Playing Sound fauled error:", error.localizedDescription)
+            print("Playing Sound failed error:", error.localizedDescription)
+            activePlayer[intent.sound.id] = nil
```

`activePlayer[intent.sound.id] = nil` can be moved after the catch block to avoid repeating.

---

actually we should probably throw an error if the sounds fails to play. in this case you can use `defer` to still nil the active player.

---

Done in 032f0c0